src/pebble.py

```diff
@@ -324,16 +326,27 @@ def change_config(  # noqa: C901 pylint: disable=too-many-branches,too-many-stat
             enable_irc_bridge(container=container, charm_state=charm_state)
             synapse.add_app_service_config_field(current_synapse_config)
             replan_irc_bridge(container=container)
-        # Push worker configuration
-        _push_synapse_config(
-            container,
-            get_worker_config(unit_number),
-            config_path=synapse.SYNAPSE_WORKER_CONFIG_PATH,
+        config_has_changed = DeepDiff(
+            existing_synapse_config,
+            current_synapse_config,
+            ignore_order=True,
+            ignore_string_case=True,
         )
-        # Push main configuration
-        _push_synapse_config(container, current_synapse_config)
-        synapse.validate_config(container=container)
-        restart_synapse(container=container, charm_state=charm_state, is_main=is_main)
+        if config_has_changed:
+            logging.info("Configuration has changed, Synapse will be restarted.")
+            logging.debug("The change is: %s", config_has_changed)
+            # Push worker configuration
+            _push_synapse_config(
+                container,
+                get_worker_config(unit_number),
+                config_path=synapse.SYNAPSE_WORKER_CONFIG_PATH,
+            )
+            # Push main configuration
+            _push_synapse_config(container, current_synapse_config)
+            synapse.validate_config(container=container)
+            restart_synapse(container=container, charm_state=charm_state, is_main=is_main)
+        else:
+            logging.info("Configuration has not changed, no action.")
```

Comment:
I think it'd be good to have a test that confirms this behaviour. Confirm that if a config-changed hook fires but configuration doesn't change the service isn't restarted, and vice versa.

Reply:
Nice, I'll add one in further PR.